File: apps/web/src/features/decisions/services/decisions.service.ts
import { createClient } from '@/lib/supabase/server'
import { complete } from '@/features/ai/services/ai-provider'
import { MomentumService } from '@/features/momentum/services/momentum.service'
import { PatternOracleService } from '@/features/pattern-oracle/services/oracle.service'

export interface DecisionOption {
  option: string
  pros: string[]
  cons: string[]
  goal_alignment: number    // 0–100
  risk_level: 'low' | 'medium' | 'high'
  effort_required: 'low' | 'medium' | 'high'
}

export interface DecisionAnalysis {
  recommendation: string          // which option + one-line reason
  options_analysis: DecisionOption[]
  key_considerations: string[]    // 3–5 things to weigh
  confidence: 'high' | 'medium' | 'low'
}

export interface Decision {
  id: string
  user_id: string
  question: string
  options: string[]
  analysis: DecisionAnalysis | null
  created_at: string
}

export class DecisionEngineService {
  static async analyze(
    userId: string,
    question: string,
    options: string[],
  ): Promise<Decision> {
    const supabase = await createClient()

    // Fetch user context to inform decision alignment with their goals/momentum
    const [
      { data: goals },
      momentumResult,
      patternsResult,
    ] = await Promise.all([
      supabase.from('goals').select('title, progress, domain').eq('user_id', userId).eq('is_active', true).limit(5),
      MomentumService.getCurrentScore(userId).catch(() => null),
      PatternOracleService.getInsights(userId).catch(() => ({ insights: [], keystone_habit: null, has_enough_data: false })),
    ])

    const goalsList = (goals ?? []).map((g) => `${g.title} (${g.progress ?? 0}% done${g.domain ? `, domain: ${g.domain}` : ''})`).join('; ')
    const topPatterns = patternsResult.insights.slice(0, 3).map((p) => `${(p as { title: string }).title}`).join(', ')
    const mom = momentumResult

    const prompt = `You are VitaMind's Decision Engine. Analyse this personal decision and provide structured, data-backed guidance.

DECISION: "${question}"

OPTIONS TO EVALUATE:
${options.map((o, i) => `${i + 1}. ${o}`).join('\n')}

USER CONTEXT:
${goalsList ? `Active goals: ${goalsList}` : 'No active goals yet'}
${mom ? `Momentum: ${mom.score}/100 | Burnout risk: ${mom.burnout_risk}/100 | Energy: ${100 - mom.burnout_risk}/100` : ''}
${topPatterns ? `Key behavioural patterns: ${topPatterns}` : ''}

Evaluate each option against the user's goals, current energy level, and behavioural patterns.

Respond in JSON only (no markdown):
{
  "recommendation": "Recommend option X because...",
  "options_analysis": [
    {
      "option": "exact option text",
      "pros": ["pro 1", "pro 2", "pro 3"],
      "cons": ["con 1", "con 2"],
      "goal_alignment": 75,
      "risk_level": "medium",
      "effort_required": "high"
    }
  ],
  "key_considerations": ["consideration 1", "consideration 2", "consideration 3"],
  "confidence": "high"
}`

    const raw = await complete({ prompt, maxTokens: 900, temperature: 0.4 })
    const parsed = JSON.parse(raw.replace(/```[\w]*\n?/g, '').replace(/```/g, '').trim()) as DecisionAnalysis

    const { data, error } = await supabase
      .from('decisions')
      .insert({
        user_id: userId,
        question,
        options,
        analysis: parsed,
      })
      .select()
      .single()

    if (error) throw error
    return data as Decision
  }

  static async getHistory(userId: string, limit = 20): Promise<Decision[]> {
    const supabase = await createClient()
    const { data, error } = await supabase
      .from('decisions')
      .select('*')
      .eq('user_id', userId)
      .order('created_at', { ascending: false })
      .limit(limit)

    if (error) throw error
    return (data ?? []) as Decision[]
  }

  static async deleteDecision(userId: string, id: string): Promise<void> {
    const supabase = await createClient()
    const { error } = await supabase
      .from('decisions')
      .delete()
      .eq('id', id)
      .eq('user_id', userId)

    if (error) throw error
  }
}

File: apps/mobile/lib/features/decisions/data/decisions_service.dart
import 'package:vitamind/core/constants/app_constants.dart';
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

class DecisionOption {
  final String option;
  final List<String> pros;
  final List<String> cons;
  final int goalAlignment;
  final String riskLevel;
  final String effortRequired;

  const DecisionOption({
    required this.option,
    required this.pros,
    required this.cons,
    required this.goalAlignment,
    required this.riskLevel,
    required this.effortRequired,
  });

  factory DecisionOption.fromMap(Map<String, dynamic> m) => DecisionOption(
        option: m['option'] as String? ?? '',
        pros: List<String>.from(m['pros'] as List? ?? []),
        cons: List<String>.from(m['cons'] as List? ?? []),
        goalAlignment: (m['goal_alignment'] as num?)?.toInt() ?? 0,
        riskLevel: m['risk_level'] as String? ?? 'medium',
        effortRequired: m['effort_required'] as String? ?? 'medium',
      );
}

class DecisionAnalysis {
  final String recommendation;
  final List<DecisionOption> optionsAnalysis;
  final List<String> keyConsiderations;
  final String confidence;

  const DecisionAnalysis({
    required this.recommendation,
    required this.optionsAnalysis,
    required this.keyConsiderations,
    required this.confidence,
  });

  factory DecisionAnalysis.fromMap(Map<String, dynamic> m) => DecisionAnalysis(
        recommendation: m['recommendation'] as String? ?? '',
        optionsAnalysis: (m['options_analysis'] as List? ?? [])
            .map((e) => DecisionOption.fromMap(e as Map<String, dynamic>))
            .toList(),
        keyConsiderations: List<String>.from(m['key_considerations'] as List? ?? []),
        confidence: m['confidence'] as String? ?? 'medium',
      );
}

class Decision {
  final String id;
  final String question;
  final List<String> options;
  final DecisionAnalysis? analysis;
  final String createdAt;

  const Decision({
    required this.id,
    required this.question,
    required this.options,
    this.analysis,
    required this.createdAt,
  });

  factory Decision.fromMap(Map<String, dynamic> m) => Decision(
        id: m['id'] as String,
        question: m['question'] as String,
        options: List<String>.from(m['options'] as List? ?? []),
        analysis: m['analysis'] != null
            ? DecisionAnalysis.fromMap(m['analysis'] as Map<String, dynamic>)
            : null,
        createdAt: m['created_at'] as String? ?? '',
      );
}

class DecisionsService {
  final String _baseUrl;
  DecisionsService()
      : _baseUrl = AppConstants.apiBaseUrl;

  Future<Map<String, String>> _authHeaders() async {
    final session = Supabase.instance.client.auth.currentSession;
    if (session == null) throw Exception('Not authenticated');
    return {
      'Authorization': 'Bearer ${session.accessToken}',
      'Content-Type': 'application/json',
    };
  }

  Future<List<Decision>> getHistory() async {
    try {
      final dio = Dio();
      final res = await dio.get(
        '$_baseUrl/api/v1/decisions',
        options: Options(headers: await _authHeaders()),
      );
      return (res.data['data'] as List? ?? [])
          .map((e) => Decision.fromMap(e as Map<String, dynamic>))
          .toList();
    } catch (e) {
      debugPrint('DecisionsService.getHistory failed: $e');
      rethrow;
    }
  }

  Future<Decision> analyze({required String question, required List<String> options}) async {
    try {
      final dio = Dio();
      final res = await dio.post(
        '$_baseUrl/api/v1/decisions',
        data: {'question': question, 'options': options},
        options: Options(headers: await _authHeaders()),
      );
      return Decision.fromMap(res.data['data'] as Map<String, dynamic>);
    } catch (e) {
      debugPrint('DecisionsService.analyze failed: $e');
      rethrow;
    }
  }

  Future<void> delete(String id) async {
    try {
      final dio = Dio();
      await dio.delete(
        '$_baseUrl/api/v1/decisions/$id',
        options: Options(headers: await _authHeaders()),
      );
    } catch (e) {
      debugPrint('DecisionsService.delete failed: $e');
      rethrow;
    }
  }
}
